### DISCUSSION TITLE
How many ModbusClientRTU and ModbusServerRTU can be run on single esp32?

### DISCUSSION THREAD

**Discussion by gytisgreitai** [eModbus/eModbus#261] [Q&A]
Hey,

so I have a Powermeter which has a modbus interface and I would like to have esp32 with eModbus in between that powermeter and two other devices: solar inverter and car charge loadpoint

For this to work I would need to have one ModbusClientRTU and two ModbusServerRTU instances running on single esp32 instance. 
Esp32 has only two cores, so I guess it would be doable if I had one ModbusClientRTU and one ModbusServerRTU - I could put them on different cores.

Currently I've implemented the ModbusClientRTU to powermeter on esphome as a platform (so that I could get convenient OTA and expose metrics to Home Assistant) with emodbus.
Current CRC check Error rate when simply reading 40 registers sequentially  every 1.5 seconds and doing nothing - is around 0%. If data push to Home Assistant is enabled the error rate is ~0.4-0.6%

Now probably the simpliest way would be to do some coding and wiring and check if it actually works, but maybe someone knows the answer beforehand - like "it won't work for sure"

Thanks!

**Comment by Miq1** [comment]
Which one is the "super client" in your setup - the power meter? And you would like to request the solar inverter and wall box from it as well? or did I get that wrong?

**Reply by gytisgreitai** [reply]
powermeter - ModbusClientRTU
solar - ModbusServerRTU
ev loadpoint - ModbusServerRTU

esp32 polls powermeter via ModbusClientRTU (`modbusClientRTUInstance.addRequest(...)`) and listens for requests from solar and ev loadpoint (`solarModbusServerRTUInstance.registerWorker(...) , evModbusServerRTUInstance.registerWorker(...)`) and answers to them with converted values that it polled from powermeter in the specific format for each of those integrations.

**Reply by Miq1** [reply]
I see. Do you have the alternative to bring both solar inverter and load point on ModbusTCP? As a TCP-to-RTU bridge the ESP32 could serve both.
On the other hand, if the frequency of requests is not too high the ESP32 might be able to serve both. The tasks will have to compete for core time, but as the serial interfaces are different, will not block each other. Did you simply try it?

**Reply by gytisgreitai** [reply]
without alternative hardware I guess I cannot to tcp-to-rtu, as neither solar nor load point support that out of the box, so would like to avoid building/buying that bridge
Solar request frequency is around ~2s. For load point - not yet sure.
Trying involves figuring out how to map different modbus request/response formats. I'm almost done with solar, but have not yet looked how to achieve that with loading point. 

Anyway, I guess this is not a hard no, so the only way is to try it out :)

Thank you for quick replies and awesome library ! Will try to report back if(when) I succeed

**Reply by Miq1** [reply]
> without alternative hardware I guess I cannot to tcp-to-rtu, as neither solar nor load point support that out of the box, so would like to avoid building/buying that bridge 

There is a ModbusBridge in the library, but you would need three ESPs then to cover all three devices.

If I was involved I would give the 2 RTU clients/ one server version a try. I do not remember we had an ESP32 using all three serials for Modbus before, so it will be interesting for us as well to learn how that works out 😉

**Reply by gytisgreitai** [reply]
ok, so that's gonna be interesting. The EV loadpoint is requesting data in a constant loop non-stop (11 requests)  so we'll have to see how much core time will be left for solar.

Now this might be a stupid question, since I'm not very familiar with modbus, but couldn't I just plug in the two esp32 into the power meter? 
The first one would have `ModbusClientRTU.addRequest` and `ModbusClientRTU.onDataHandler` and would be used by EV loadpoint
The other one would have only `ModbusClientRTU.onDataHandler` and be dependent on the first one `.addRequest` and would be used by solar? The modbus data simply flows on the wire, and you can just intercept it, or no?

**Reply by Miq1** [reply]
You could setup the second ESP32 as a listener device. For RTU servers there is a special ``registerSniffer()`` method to define a callback that will get any message from the bus, regardless if request or response. This sniffer can attach to the RTU bus between the other ESP32 and the power meter, see everything, but is not able to send requests on that bus. It would be your task in the callback to recognize the responses and extract and set aside the data from it. The other side would be a regular server again that will serve the requests from the solar inverter with these copied data.
Interesting concept. 😉

**Reply by gytisgreitai** [reply]
and hopefully last midly related question. How do I send float (e.g. 100.01) as 4 modbus registers? Should I use a double? Floats if used via `modbusMessage.add()` are two registers?

![image](https://user-images.githubusercontent.com/1690228/213907727-fe0fe287-3206-46c5-87da-fb699f7c3dc7.png)

**Reply by Miq1** [reply]
Yes, ``float`` is a ``float32``. If you need 64 bits, use ``double`` instead.

**Reply by Miq1** [reply]
By the way: I would read the table row above differently as ``uint64_t`` being sent that needs to be divided by 100.0 to get the ``float`` kWh.

**Reply by gytisgreitai** [reply]
this is the meter that i would be “faking” ( that is writing not reading) so does that also apply in reverse? Use `uint64_t` in `message.add` instead of `double`?

**Reply by Miq1** [reply]
Yeah, why not. Give it a try.

**Reply by gytisgreitai** [reply]
So thanks for your help. Finally understood that there are "two" types of floats. The IEEE754 and simply integer with fixed precision.

I am surprised, but it actually works. The full solution is running for ~74 hours, seems to be quite ok. 
The CRC error rate is hovering around 0.5-1%.

The only remaining issue for which I can probably add work around, is that sometimes I get strange fluctuations with total kWh, e.g. 846->5000->846 or 1105->9350>1105

**Reply by Miq1** [reply]
👍🏻
